pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,245 @@
+	delegatedidentityv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/agent/delegatedidentity/v1"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/spiffe/spire/pkg/agent/api/rpccontext"
+	workload_attestor "github.com/spiffe/spire/pkg/agent/attestor/workload"
+	"github.com/spiffe/spire/pkg/agent/endpoints"
+	"github.com/spiffe/spire/pkg/agent/manager"
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/proto/spire/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RegisterService registers the delegated identity service on the provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	delegatedidentityv1.RegisterDelegatedIdentityServer(s, service)
+}
+
+type attestor interface {
+	Attest(ctx context.Context) ([]*common.Selector, error)
+}
+
+type Config struct {
+	Log                 logrus.FieldLogger
+	Manager             manager.Manager
+	Attestor            workload_attestor.Attestor
+	AuthorizedDelegates []string
+}
+
+func New(config Config) *Service {
+	AuthorizedDelegates := map[string]bool{}
+
+	for _, delegate := range config.AuthorizedDelegates {
+		AuthorizedDelegates[delegate] = true
+	}
+
+	return &Service{
+		manager:             config.Manager,
+		attestor:            endpoints.PeerTrackerAttestor{Attestor: config.Attestor},
+		AuthorizedDelegates: AuthorizedDelegates,
+	}
+}
+
+// Service implements the delegated identity server
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor attestor
+
+	// SPIFFE IDs of delegates that are authorized to use this API
+	AuthorizedDelegates map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized delegates map.
+func (s *Service) isCallerAuthorized(ctx context.Context, log logrus.FieldLogger) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return false, status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+	numRegisteredIDs := len(identities)
+
+	if numRegisteredIDs == 0 {
+		log.Error("no identity issued")
+		return false, status.Error(codes.PermissionDenied, "no identity issued")
+	}
+
+	for _, identity := range identities {
+		if _, ok := s.AuthorizedDelegates[identity.Entry.SpiffeId]; ok {
+			return true, nil
+		}
+	}
+
+	// caller has identity associeted with but none is authorized
+	log.WithFields(logrus.Fields{
+		"num_registered_ids": numRegisteredIDs,
+		"default_id":         identities[0].Entry.SpiffeId,
+	}).Error("Permission denied; caller not configured as an authorized delegate.")
+
+	return false, nil
```

Comment:
@amartinezfayo I have inserted theses lines here inside isCallerAuthorized():
```
log.WithFields(logrus.Fields{
		"num_registered_ids": numRegisteredIDs,
		"default_id":         identities[0].Entry.SpiffeId,
	}).Error("Permission denied; caller not configured as an authorized delegate.")
```

What do you think about take the following line about status.Error and put together here and remove it from SubscribeToX509SVIDs()? Doing so, isn't necessary to return `authorized`.
```
	if !authorized {
		return status.Error(codes.PermissionDenied, "caller not configured as an authorized delegate")
	}
```

Reply:
That would be a nice simplification of the logic!

Reply:
hey @amartinezfayo, so, inside `SubscribeToX509SVIDs` and `SubscribeToX509Bundles` end up like this:
https://github.com/spiffe/spire/blob/954571c17293086416d424f0819bb4d84ae2a91a/pkg/agent/api/delegatedidentity/v1/service.go#L101-L107

Even if we didn't use `authorized` variable anymore inside both functions, should we return a bool from `isCallerAuthorized` for future use? Or just `err` is enough in this case?

Reply:
We can just have
```
	if err := s.isCallerAuthorized(ctx, log); err != nil {
		return err
	}
```
I would just remove the bool returned, since it's not needed.

Reply:
Perfect! tks

Reply:
In the last commit a change the return of `isCallerAuthorized`, to return the caller's selectors to cache it. What do you think?

https://github.com/spiffe/spire/blob/7bb7b88145a781918557590a271a8d4e5ca43e49/pkg/agent/api/delegatedidentity/v1/service.go#L105-L109